pkg/streamingpromql/engine_fuzz_test.go

```diff
@@ -0,0 +1,234 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/tree/main/promql/engine_test.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package streamingpromql
+
+import (
+	"context"
+	"math"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/util/teststorage"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/streamingpromql/testutils"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/promqltest"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	types.EnableManglingReturnedSlices = true
+	parser.ExperimentalDurationExpr = true
+	parser.EnableExperimentalFunctions = true
+}
+
+// we ignore these as they include the string index of where the invalid value is in the query
+// seems like prometheus does a trim() somewhere but MQE does not, so we end up with different char index's referenced in the warning
+var ignorePrefix = []string{"PromQL warning: quantile value should be between 0 and 1", "PromQL info: metric might not be a counter, name does not end in"}
+
+// Replace series Metric which are nil with an empty Labels{}
+func fixUpEmptyLabels(r *promql.Result) {
+	if r == nil || r.Value == nil {
+		return
+	}
+	switch r.Value.Type() {
+	case parser.ValueTypeMatrix:
+		matrix, _ := r.Matrix()
+		for i := 0; i < len(matrix); i++ {
+			if matrix[i].Metric == nil {
+				matrix[i].Metric = labels.Labels{}
+			}
+		}
+	}
+}
+
+func DoComparisonQuery(t *testing.T, asInstantQuery bool, query string, prometheusEngine *promql.Engine, mimirEngine *Engine, testStorage *teststorage.TestStorage) {
+
+	var prom, mimir *promql.Result
+	var promErr, mimiErr error
+	var promQry, mimiQry promql.Query
+
+	if asInstantQuery {
+
+		ts := time.Now()
+
+		promQry, promErr = prometheusEngine.NewInstantQuery(context.Background(), testStorage, nil, query, ts)
+		mimiQry, mimiErr = mimirEngine.NewInstantQuery(context.Background(), testStorage, nil, query, ts)
+
+	} else {
+
+		start := time.Now()
+		end := time.Now().Add(time.Minute)
+		step := time.Minute
+
+		promQry, promErr = prometheusEngine.NewRangeQuery(context.Background(), testStorage, nil, query, start, end, step)
+		mimiQry, mimiErr = mimirEngine.NewRangeQuery(context.Background(), testStorage, nil, query, start, end, step)
+	}
+
+	if promErr == nil {
+		defer promQry.Close()
+		prom = promQry.Exec(context.Background())
+	}
+
+	if mimiErr == nil {
+		defer mimiQry.Close()
+		mimir = mimiQry.Exec(context.Background())
+	}
+
+	if promErr == nil {
```

Comment:
Would it be possible to simplify some of these conditions by starting by checking `if promErr != nil`? Then, if that is not the case, then we can `require.NoError(t, mimirErr)` and eliminate most of the `if`s